accepted/system-range/system-range.md

```diff
@@ -0,0 +1,170 @@
+# System.Range
+
+**PM** [Immo landwerth](https://github.com/terrajobst)
+
+The C# team is considering providing language syntax for specifying ranges to
+represent parts of a string, array, or span. In order to make this work, the
+language has to be bound to a .NET type, which we call `System.Range`.
+
+## Scenarios and User Experience
+
+### Constructing and enumerating ranges
+
+```csharp
+// Ranges are constructed with an index and a length
+
+var fiveToTen = new Range(index: 5, length: 5);
+
+// By using a factory method, they can also be constructed from two indices.
+// The upper bound is considered inclusive.
+
+var alsoFiveToTen = Range.Between(5, 10);
+
+// Ranges are enumerable so we can use it to print the values from 1 to (and
+// including) 3:
+
+foreach (var value in Range.Between(1, 3))
+    Console.WriteLine(value);
+```
+
+### Dealing with open ended ranges
+
+```csharp
+// Ranges can be open ended:
+
+var fiveToEnd = Range.From(5); // Missing upper bound
+var startToTen = Range.To(10); // Missing lower bound
+var everything = Range.All;    // Missing upper and lower bound
+
+// Asking for Index, Length or enumerating an open range will throw
+// NotSupportedException. The consumer of the range is expected to 
+// close it over the particular data:
+
+void Print(int[] data, Range range)
+{
+    if (range.IsOpen)
+        range = range.Close(0, data.Length);
+
+    foreach (var index in range)
+        Console.WriteLine(data[index]);
+}
+
+var data = new [] { 'a', 'b', 'c' };
+Print(data, Range.From(2)); // Prints c
+Print(data, Range.To(1));   // Prints a, b
+Print(data, Range.All);     // Prints a, b, c
+```
+
+### Interplay with C# syntax
+
+The syntax for ranges isn't finalized and is actively worked on. While this is
+part of a separate design group it makes sense to look at some samples to see
+how this BCL type will play with it:
+
+```
+var fiveToTen = 5..10; // Equivalent to Range.Between(5, 10)
+var fiveToRest = 5..;  // Equivalent to Range.From(5)
+var startToTen = ..10; // Equivalent to Range.To(10)
+var everything = ..;   // Equivalent to Range.All
+
+```
+
+### Using ranges in APIs
+
+Ranges provide a concise way to provide an `index` and a `length`. This allows
+for making common operations more compact and less confusing at the call side.
+For example, consider this version of the indexer for `String`:
+
+```csharp
+partial class String
+{
+    public string this[Range range] => Substring(range.Index, range.Length);
+}
+```
+
+This allows creating substrings simply by using the indexer:
+
+```csharp
+var helloWorld = "Hello, World!";
+var hello = helloWorld[..5];
+var world = helloWorld[7..5];
```

Comment:
Should it be 5..7?

Reply:
Could you add as a comment what is expected result is for each line, it'll be easier to understand your thinking.

Reply:
That's what the variable names are for ;-)

---
In reply to: [163118819](https://github.com/dotnet/designs/pull/30#discussion_r163118819) [](ancestors = 163118819)